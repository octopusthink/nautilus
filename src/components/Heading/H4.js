import { css } from '@emotion/core';
import styled from '@emotion/styled';
import React from 'react';

import Heading from './';

// We define `<H4>` for developer convenience.
export const Heading4 = (props) => {
  return <Heading level={4} {...props} />;
};

Heading4.defaultProps = {
  children: undefined,
};

Heading4.propTypes = { ...Heading.PropTypes, level: undefined };

export default styled(Heading4)``;
