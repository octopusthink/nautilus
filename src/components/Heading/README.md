Headings are short pieces of text used to organise and structure a page's content.

They provide both a semantic and a presentational role, guiding users (as well as search engines and other bots) through the page content both visually (through hierarchy) and structurally (through markup). They are especially important for users of assistive technology‚Äîheading allow quick navigation throughout a page.

```jsx
<Heading>This just in: Scotland is rainy üè¥Û†ÅßÛ†Å¢Û†Å≥Û†Å£Û†Å¥Û†Åøüå¶</Heading>
```

## Usage

Use a `<Heading>` to introduce a block of text or other content. Headings act as titles for sections of content, and are almost always followed by more detailed information.

Don't use a `<Heading>` for:
- The primary title of a page. Use a `<PageTitle>` instead.
- A subtitle underneath a primary title. Use a `<Subtitle>` instead.
- Long blocks of text intended to span more than two lines. Use a [`<Paragraph>`](../#/Function/Components/Paragraph) instead.
- A primary link that stands by itself. Use a [`<Button>`](../#/Function/Components/Button) instead.

## Appearance

- Use a consistent colour and style for headings across the site.
- Use variations in size to establish hierarchy in headings.
- The space between a heading and the content it titles should be notably smaller than the space between the heading and its preceding content. This helps establish a relationship by ensuring that headings are in close proximity to the content they describe.
- Headings shouldn't span longer than two lines.
- Headings should be visibly larger and more prominent than body copy, and should be more tightly spaced between lines.

## Variations

Headings are provided in three different sizes. You should **always use headings sequentially** (a medium heading would be nested underneath a large heading).

```jsx
<Heading level={2}>Breaking news üì∞</Heading>
<Heading level={3}>Kittens are better than dogs üòº</Heading>
<Heading level={4}>More at eleven üïö</Heading>
```

The `large` heading then uses an `<h2>`, the `medium` heading uses `<h3>`, and the `small` heading uses `<h4>`. You should not break this levelling structure; see the Accessibility section for more details.

## Voice & Tone

- Headings should use Sentence case, not Title Case.
- Headings should use a full stop (period) only when they are written as full sentences.
- Keep headings as short as possible. Only use a full sentence for a heading when necessary for stylistic purposes.
- Use headings to separate and organise long blocks of content.
- Headings should reflect and title the content that follows them.

## Accessibility

For accessibility reasons, it's important that `<h>` tags follow proper semantic order.

- Never skip a heading level (ie, going from `<h2>` to `<h4>`).
- Don't traverse heading levels (ie, nesting an `<h2>` under an `<h4>`). Headings should be used sequentially to provide further separation inside an existing section.
- Only use a single `<h1>` tag per page. (This should be your `<PageTitle>`.)
- Don't write headings in ALL CAPS. The variation of font, size, and/or weight should be a sufficient indicator of hierarchy.
