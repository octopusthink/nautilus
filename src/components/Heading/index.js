import { css } from '@emotion/core';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

import { useTheme } from 'themes';
import { headingLarge, headingMedium, headingSmall } from 'themes/mixins';

const LARGE = 'large';
const MEDIUM = 'medium';
const SMALL = 'small';

const HeadingLevels = {
  [LARGE]: 2,
  [MEDIUM]: 3,
  [SMALL]: 4,
};

export const Heading = ({
  children,
  large,
  level,
  medium,
  small,
  ...otherProps
}) => {
  invariant(
    [large, medium, small].filter((prop) => prop).length <= 1,
    '<Heading> can only be one of large/medium/small. Please only set one.',
  );

  // Headings are large by default.
  let size = LARGE;
  if (medium) {
    size = MEDIUM;
  }
  if (small) {
    size = SMALL;
  }

  const HeadingElement = `h${HeadingLevels[size]}`;
  const theme = useTheme();

  return (
    <HeadingElement
      css={css`
        margin: 0 0 ${theme.spacing.margin.m};

        ${large &&
          css`
            ${css(headingLarge(theme))}
          `}

        ${medium &&
          css`
            ${css(headingMedium(theme))};
          `}

        ${small &&
          css`
            ${css(headingSmall(theme))}
          `}
      `}
      {...otherProps}
    >
      {children}
    </HeadingElement>
  );
};

Heading.defaultProps = {
  children: undefined,
  level: 2,
  large: false,
  medium: false,
  small: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Semantic hierarchy level of the `<h>` element in the markup (ex: `<h3>`). If no size is provided, a `level={2}` outputs a `large` heading; `level={3}` a `medium` heading, and `level={4}` a `small` heading. */
  level: PropTypes.oneOf(Object.keys(HeadingLevels)).isRequired,
  /** Make this heading visually large in size. */
  large: PropTypes.bool,
  /** Make this heading visually medium in size. */
  medium: PropTypes.bool,
  /** Make this heading visually small in size. */
  small: PropTypes.bool,
};

export default Heading;
