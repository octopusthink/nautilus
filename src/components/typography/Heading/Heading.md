## Available Sizes
### XXLarge Heading
```js
<Heading level={2} size="xxlarge">This is a heading</Heading>
```

### XLarge Heading
```js
<Heading level={2} size="xlarge">This is a heading</Heading>
```

### Large Heading
```js
<Heading level={2} size="large">This is a heading</Heading>
```

### Medium Heading
```js
<Heading level={2}>This is a heading</Heading>
```

### Small Heading
```js
<Heading level={2} size="small">This is a heading</Heading>
```

### XSmall Heading
```js
<Heading level={2} size="xsmall">This is a heading</Heading>
```

### XXSmall Heading
```js
<Heading level={2} size="xxsmall">This is a heading</Heading>
```

## Colors
### Headings with custom colors
```js
<Heading level={2} color="light">This is a heading</Heading>
<Heading level={2} color="medium">This is a heading</Heading>
<Heading level={2} color="dark">This is a heading</Heading>
<Heading level={2} color="accent">This is a heading</Heading>
<Heading level={2} color="danger">This is a heading</Heading>
<Heading level={2} color="success">This is a heading</Heading>
```
