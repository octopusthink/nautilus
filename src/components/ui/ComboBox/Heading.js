import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { toUnits } from '../../../styles';
import { useTheme } from '../../../themes';
import ListItem from '../List/Item';

export const ComponentClassName = 'Nautilus-ComboBoxHeading';

export const Heading = (props) => {
  const { children, className, unstyled, ...otherProps } = props;

  const theme = useTheme();

  return (
    <ListItem
      className={classnames(ComponentClassName, className)}
      css={
        unstyled
          ? undefined
          : css`
              background: ${theme.colors.text.default};
              color: ${theme.colors.text.inverse};
              line-height: 2.4rem;
              padding: ${toUnits(theme.spacing.padding.medium)};
              width: 100%;
            `
      }
      unstyled
      {...otherProps}
    >
      {children}
    </ListItem>
  );
};

Heading.defaultProps = {
  children: undefined,
  className: undefined,
  unstyled: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Heading;

export default Heading;
