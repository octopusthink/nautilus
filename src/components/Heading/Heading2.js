import { css } from '@emotion/core';
import styled from '@emotion/styled';
import React from 'react';

import Heading from './';

// We define `<Heading2>` for developer convenience.
export const Heading2 = (props) => {
  return <Heading level={2} {...props} />;
};

Heading2.defaultProps = {
  children: undefined,
};

Heading2.propTypes = { ...Heading.PropTypes, level: undefined };

export default styled(Heading2)``;
