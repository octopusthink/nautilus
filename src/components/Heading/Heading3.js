import { css } from '@emotion/core';
import styled from '@emotion/styled';
import React from 'react';

import Heading from './';

// We define `<Heading3>` for developer convenience.
export const Heading3 = (props) => {
  return <Heading level={3} {...props} />;
};

Heading3.defaultProps = {
  children: undefined,
};

Heading3.propTypes = { ...Heading.PropTypes, level: undefined };

export default styled(Heading3)``;
