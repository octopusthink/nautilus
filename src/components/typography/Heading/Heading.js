import PropTypes from 'prop-types';
import React from 'react';

import { headingSizes, headingColors, headingElements } from './Heading.styles';

export const Heading = ({ children, color, level, size, ...others }) => {
  const HeadingElement = headingElements[`h${level}`];
  return (
    <HeadingElement
      color={color}
      size={size}
      {...others}
    >
      {children}
    </HeadingElement>
  );
};

export const HeadingColors = Object.keys(headingColors);
export const HeadingLevels = [1, 2, 3, 4, 5, 6];
export const HeadingSizes = Object.keys(headingSizes);

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Color of the Heading. Options are 'dark', 'medium', 'light', 'accent', 'danger', and 'success' */
  color: PropTypes.oneOf(HeadingColors),
  /** Hierarchy level of the Heading (ex: h3) */
  level: PropTypes.oneOf(HeadingLevels),
  /** Size of the Heading. Options are 'xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge' */
  size: PropTypes.oneOf(HeadingSizes),
};

Heading.defaultProps = {
  color: 'dark',
  level: 2,
  size: 'medium',
};
