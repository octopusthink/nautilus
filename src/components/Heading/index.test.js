import React from 'react';

import { render } from 'utils/testing';

import Heading from '.';

describe('Heading', () => {
  it('should render an <h2> by default', () => {
    const { container } = render(<Heading>Hello</Heading>);

    expect(container.firstChild.tagName).toEqual('H2');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Heading>
        <div data-testid="child" />
      </Heading>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should render an <h2> tag when the `large` prop is set', () => {
    const { container } = render(<Heading large>Big heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H2');
  });

  it('should render an <h3> tag when the `medium` prop is set', () => {
    const { container } = render(<Heading medium>Regular heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H3');
  });

  it('should render an <h4> tag when the `small` prop is set', () => {
    const { container } = render(<Heading small>Tiny heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H4');
  });

  it('should output large CSS when the `large` prop is set', () => {
    const { container } = render(<Heading large>Big heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should output medium CSS when the `medium` prop is set', () => {
    const { container } = render(<Heading medium>So-so heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should output small CSS when the `small` prop is set', () => {
    const { container } = render(<Heading small>Tiny heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should only allow one of large/medium/small props', () => {
    const errorText =
      '<Heading> can only be one of large/medium/small. Please only set one.';

    expect(() => {
      render(
        <Heading large medium small>
          Big and small
        </Heading>,
      );
    }).toThrow(errorText);

    expect(() => {
      render(
        <Heading large small>
          Big and small
        </Heading>,
      );
    }).toThrow(errorText);

    expect(() => {
      render(
        <Heading medium small>
          Big and small
        </Heading>,
      );
    }).toThrow(errorText);

    expect(() => {
      render(
        <Heading large medium>
          Big and small
        </Heading>,
      );
    }).toThrow(errorText);
  });
});
