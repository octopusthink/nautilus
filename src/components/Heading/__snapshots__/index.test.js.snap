// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should output large CSS when the \`large\` prop is set 1`] = `
.emotion-0 {
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 5.5rem;
  line-height: 1.018181818181818;
}

<h2
  class="emotion-0"
>
  Big heading
</h2>
`;

exports[`Heading should output medium CSS when the \`medium\` prop is set 1`] = `
.emotion-0 {
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 4.4rem;
  line-height: 1;
}

<h3
  class="emotion-0"
>
  So-so heading
</h3>
`;

exports[`Heading should output small CSS when the \`small\` prop is set 1`] = `
.emotion-0 {
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 3.5rem;
  line-height: 1.0285714285714285;
}

<h4
  class="emotion-0"
>
  Tiny heading
</h4>
`;
